doc/source/status.rst

```diff
@@ -1,36 +1,243 @@
+:obj:`~ophyd.utils.StatusTimeoutError` will be raised. If a timeout given to
+:obj:`~ophyd.StatusBase.wait` expires before any of these things happen,
+:obj:`~ophyd.utils.WaitTimeoutError` will be raised.
+
+The method :obj:`~ophyd.StatusBase.exception` behaves similarly to
+:obj:`~ophyd.StatusBase.wait`; the only difference is that if the Status is marked as
+failed or the Status' own timeout expires it *returns* the exception rather
+than *raising* it. Both return ``None`` if the Status finishes successfully,
+and both raise :obj:`~ophyd.utils.WaitTimeoutError` if the given timeout expires
+before the Status completes or times out.
+
+Alternatively, the recipient of the Status object can ask to be notified of
+completion asynchronously by adding a callback. The callback will be called
+when the Status is marked as complete or its timeout has expired. (If no
+timeout was given, the callback might never be called. This is why providing a
+timeout is strongly recommended.)
+
+.. code:: python
+
+   def callback(status):
+       print(f"{status} is done")
+
+   status.add_callback(callback)
+
+Callbacks may be added at any time. Until the Status completes, it holds a hard
+reference to each callback in a list, ``status.callbacks``. The list is cleared
+when the callback completes. Any callbacks added to a Status object *after*
+completion will be called immediately, and no reference will be held.
+
+Each callback is passed to the Status object as an argument, and it can use
+this to distinguish success from failure.
+
+.. code:: python
+
+   def callback(status):
+       error = status.exception()
+       if error is None:
+           print(f"{status} has completed successfully.")
+       else:
+           print(f"{status} has failed with error {error}.")
+
+SubscriptionStatus
+------------------
+
+The :class:`~ophyd.status.SubscriptionStatus` is a special Status object that
+correctly and succinctly handles a common use case, wherein the Status object
+is marked finished based on some ophyd event. It reduces this:
+
+.. code:: python
+
+   from ophyd import Device, Component, DeviceStatus
+
+   class MyToyDetector(Device):
+       ...
+       # When set to 1, acquires, and then goes back to 0.
+       acquire = Component(...)
+
+       def trigger(self):
+           def check_value(old_value, value, **kwargs):
+               if old_value == 1 and value == 0:
+                   status.set_finished()
+                   # Clear the subscription.
+                   sself.acquire.clear_sub(check_value)
+
+           status = DeviceStatus(self.acquire)
+           self.acquire.subscribe(check_value)
+           self.acquire.set(1)
+           return status
+
+to this:
+
+.. code:: python
+
+   from ophyd import Device, Component, SubscriptionStatus
+
+   class MyToyDetector(Device):
+       ...
+       # When set to 1, acquires, and then goes back to 0.
+       acquire = Component(...)
+
+       def trigger(self):
```

Comment:
This does not match what the code expects.   It looks like the code expect this callback to return True/False and then calls `self._finished` for you.

Reply:
Thanks, fixed.